File: packages/tabtin-chat-client/src/managers/ContextManager.ts
import { HttpClient } from '../core/http-client'
import { ContextResponse, UpdateContextRequest } from '../types'

/**
 * 上下文管理器
 * 负责会话上下文的获取、更新、清除操作
 */
export class ContextManager {
  constructor(private http: HttpClient) {}

  /**
   * 获取会话上下文
   * @param sessionId - 会话ID
   * @returns 上下文响应
   */
  async get(sessionId: string): Promise<ContextResponse> {
    return this.http.get<ContextResponse>(`/sessions/${sessionId}/context`)
  }

  /**
   * 更新会话上下文
   * @param sessionId - 会话ID
   * @param updates - 更新的上下文字段
   * @returns 更新后的上下文响应
   */
  async update(
    sessionId: string,
    updates: UpdateContextRequest
  ): Promise<ContextResponse> {
    return this.http.put<ContextResponse>(`/sessions/${sessionId}/context`, updates)
  }

  /**
   * 清除当前上下文
   * @param sessionId - 会话ID
   * @returns 删除成功消息
   */
  async clear(sessionId: string): Promise<{ message: string }> {
    return this.http.delete<{ message: string }>(`/sessions/${sessionId}/context`)
  }

  /**
   * 设置当前 Space
   * @param sessionId - 会话ID
   * @param spaceId - Space ID，传入 null 可清空
   */
  async setCurrentSpace(sessionId: string, spaceId: string | null): Promise<ContextResponse> {
    return this.update(sessionId, { current_space_id: spaceId })
  }

  /**
   * 设置当前协作 Project；资源宿主请使用 setCurrentSpace。
   */
  async setCurrentProject(sessionId: string, projectId: string | null): Promise<ContextResponse> {
    return this.update(sessionId, { current_project_id: projectId })
  }

  /**
   * 设置当前表格
   * @param sessionId - 会话ID
   * @param tableId - 表格ID，传入 null 可清空
   */
  async setCurrentTable(sessionId: string, tableId: string | null): Promise<ContextResponse> {
    return this.update(sessionId, { current_table_id: tableId })
  }

  /**
   * 设置当前视图
   * @param sessionId - 会话ID
   * @param viewId - 视图ID，传入 null 可清空
   */
  async setCurrentView(sessionId: string, viewId: string | null): Promise<ContextResponse> {
    return this.update(sessionId, { current_view_id: viewId })
  }
}

File: apps/tabtin_django/apps/chat/conversation/services/context_manager.py
"""
Context Manager - 上下文管理服务

负责管理会话上下文，帮助AI理解当前对话环境
"""

from typing import Dict, List, Optional

from ..models import ChatSession, ChatContext


class ContextManager:
    """
    上下文管理器

    功能：
    - 获取/更新会话上下文
    - 管理当前关注对象（space/table/view）
    - 管理最近访问历史
    - 为 Agent 提供上下文信息
    """

    @staticmethod
    def get_or_create_context(session_id: str, *, user_id: Optional[str]) -> ChatContext:
        """获取或创建会话上下文。

        user_id 为 keyword-only 必填参数。外部 API 调用方必须传入当前
        用户 ID 以校验会话归属；内部信任路径（如 Agent 引擎）传入从
        state 获取的 user_id（可能为 None，此时跳过归属校验）。
        """
        session = ChatSession.objects.filter(id=session_id).first()
        if not session:
            raise ValueError(f"ChatSession {session_id} not found")
        if user_id is not None and str(session.user_id) != str(user_id):
            from django.core.exceptions import PermissionDenied
            raise PermissionDenied("ChatSession does not belong to this user")
        context, created = ChatContext.objects.get_or_create(session=session)
        return context

    @staticmethod
    def get_context_summary(
        session_id: str, *, user_id: Optional[str] = None
    ) -> Dict:
        """获取上下文摘要，用于传递给 Agent 引擎"""
        context = ContextManager.get_or_create_context(session_id, user_id=user_id)

        return {
            "current_space_id": context.current_space_id,
            "current_project_id": str(context.current_project_id or context.session.project_id or ""),
            "current_table_id": context.current_table_id,
            "current_view_id": context.current_view_id,
            "recent_spaces": context.recent_spaces,
            "recent_tables": context.recent_tables,
            "recent_views": context.recent_views,
            "context_data": context.context_data,
            "conversation_summary": context.context_data.get("conversation_summary")
        }

    @staticmethod
    def set_current_table(
        session_id: str, table_id: str,
        space_id: str = None, *, user_id: Optional[str] = None,
    ) -> ChatContext:
        """设置当前表格"""
        context = ContextManager.get_or_create_context(session_id, user_id=user_id)
        context.set_current_table(table_id, space_id)
        return context

    @staticmethod
    def set_current_space(
        session_id: str, space_id: str, *, user_id: Optional[str] = None,
    ) -> ChatContext:
        """设置当前智能体空间"""
        context = ContextManager.get_or_create_context(session_id, user_id=user_id)
        context.current_space_id = space_id
        context.add_recent_space(space_id)
        return context

    @staticmethod
    def clear_current(session_id: str, *, user_id: Optional[str] = None) -> ChatContext:
        """清除当前上下文"""
        context = ContextManager.get_or_create_context(session_id, user_id=user_id)
        context.clear_current()
        return context

    @staticmethod
    def update_context_data(
        session_id: str, key: str, value, *, user_id: Optional[str] = None
    ) -> ChatContext:
        """更新自定义上下文数据"""
        context = ContextManager.get_or_create_context(session_id, user_id=user_id)
        context.context_data[key] = value
        context.save(update_fields=['context_data', 'updated_at'])
        return context

    @staticmethod
    def get_context_for_prompt(
        session_id: str, *, user_id: Optional[str] = None,
    ) -> str:
        """生成用于 Prompt 的上下文描述，可直接插入到 system prompt 中。"""
        context_summary = ContextManager.get_context_summary(session_id, user_id=user_id)

        parts = ["【当前上下文】"]

        if context_summary["current_table_id"]:
            parts.append(f"- 当前表格: {context_summary['current_table_id']}")

        if context_summary["current_space_id"]:
            parts.append(f"- 当前智能体空间: {context_summary['current_space_id']}")

        if context_summary["current_project_id"]:
            parts.append(f"- 当前协作 Project: {context_summary['current_project_id']}")

        if context_summary["recent_tables"]:
            recent = ", ".join(context_summary["recent_tables"][:3])
            parts.append(f"- 最近表格: {recent}")

        conversation_summary = context_summary.get("conversation_summary") if context_summary else None
        if conversation_summary:
            parts.append("【对话摘要】")
            parts.append(conversation_summary)

        if len(parts) == 1:
            return ""

        return "\n".join(parts)
